test/distributed/pipelining/test_schedule.py

```diff
@@ -52,6 +52,41 @@ def setUpClass(cls):
         dev_id = cls.rank % torch.cuda.device_count()
         cls.device = torch.device(f"cuda:{dev_id}")
 
+    @requires_nccl()
+    @skip_but_pass_in_sandcastle_if(not TEST_MULTIGPU, "NCCL test requires 2+ GPUs")
+    @parametrize("ScheduleClass", [ScheduleGPipe])
+    def test_forward_only(self, ScheduleClass):
+        mod = MultiMLP(d_hid, n_layers=self.world_size)
+        mod.to(self.device)
+
+        x = torch.randn(batch_size, d_hid, device=self.device)
+
+        # Create a pipeline
+        chunks = 4
+        split_spec = mod.split_spec if hasattr(mod, "split_spec") else None
+        pipe = pipeline(
+            mod,
+            chunks,
+            example_args=(x,),
+            split_spec=split_spec,
+        )
+
+        stage = PipelineStage(
+            pipe,
+            self.rank,
+            device=self.device,
+        )
+
+        # Attach to a schedule
+        schedule = ScheduleClass(stage, chunks)
```

Comment:
hmm. API concern-
it does not feel well-thought-out to me to offer inference via 'ScheduleClass' minus 'loss'

Why?
- you have to construct a separate schedule instance for 'inference' if you already had one for training

I could see either of 2 alternatives being cleaner
1) Schedule base offers a 'step_inference' api.  All schedules inherit this. It essentially ignores the real schedule logic and runs some basic inference schedule.  (maybe 2 bases, one for simple, one for looped)
2) we offer a separate 'InferenceScheduleSimple' and 'InferenceScheduleMulti' class.  Then we don't ask users to to create a second 'GPipe' schedule and signal that its for inference just by omitting loss

I think (1) is superior. wdyt?

Reply:
Thanks for the comments. Good thoughts.

First thing first, re this API: 
```
schedule = ScheduleClass(stage, chunks)
```
The reason it takes no loss function is that user has no loss function to pass.
Then, whether ScheduleClass should be `ScheduleGPipe` or `ScheduleInference` is up for discussion.

Moving onto the case of train-eval switching, `step_inference` API sounds acceptable to me (or `step_eval`). 
Or, if we don't want to handle the tiny difference between `no_grad`, `eval` and `inference`, then we can leave it to torch's existing context manager:
``` 
with torch.inference_mode():
  out = schedule.step(x)
```
We can internally detect whether `inference_mode` is on, and skip loss computation and backwards.

Reply:
i am not sure what the tradeoffs are with torch.inference_mode vs making an explicit api.

i guess we could pretty easily add inference_mode support later but we need to implement the basic step_eval thing first.  Later our 'step' could route to step_eval in inference mode if we think its worth it?

Reply:
Yeah, maybe they can compose, like:
```
with torch.inference_mode():
  out = schedule.step_eval(x)
```

Reply:
What would user expect us to do in `step_eval()` in addition to regular forward calling?
Since user has no more access to "module", would they expect this API to call `module.eval()` for them?
Is there anything else?